### PR TITLE
Server options as constructor argument

### PR DESCRIPTION
**Pull request by basz**
Before I continue, do you think approach is correct?

IMO this is ok... But looking at the test, it is possible to change behaviour by mutating the ServerOptions. When we do this everywhere that might lead to unexpected results?

### CODE REVIEW

**Comment by basz** [comment]
@ojhaujjwal 
> Why not just pass that option as a constructor argument?

This ok by you?

cc/@bakura10

**Comment by ojhaujjwal** [comment]
@basz thanks for tagging. looks great :+1:

**Comment by danizord** [comment]
@basz since we're breaking BC anyway and now `ServerOptions` instance is stored everywhere, I think it would be nice to make `ServerOptions` immutable to avoid side effects.

**Comment by ojhaujjwal** [comment]
@danizord agreed. setters are a big NO.

**Comment by bakura10** [comment]
+1

**Comment by basz** [comment]
• done •

File: src/Options/ServerOptions.php [hunk 1]
```diff
@@ -92,7 +92,7 @@ public function __construct($options = null)
      *
      * @param  array $options
      */
-    public function setFromArray(array $options)
+    protected function setFromArray(array $options)
```
Comment (bakura10) [inline src/Options/ServerOptions.php]: We've done this with @danizord on our project lately: you could create a private constructor, and using a static "fromArray" method that construct the object. This way, all the setter could be removed, which will ensure true immutability.

Comment (basz) [inline src/Options/ServerOptions.php]: okey, ... how do you ensure passed values are of the correct type?

```
setRevokeRotatedRefreshTokens(bool $revokeRotatedRefreshTokens)
```
only accepts bool

**Comment by basz** [comment]
@bakura10 this what yo mean?

plus a fail; but I don't see why; tests/src/Options/ServerOptionsTest.php:64

File: src/Options/ServerOptions.php (hunk 1 above)
Comment (danizord) [inline src/Options/ServerOptions.php]: Also, the named constructor should call the private constructor, which sets the properties. This way you always pass through type hints.

**Comment by basz** [comment]
@bakura10 @danizord I'm not sure I like the current implemetation of ServerOptions very much.

Main point; It introduces a dependency for a problem that isn't there. 
Bloats; The class is lightweight enough for a few setters, which when private still ensure immutability. (class is also marked final). Now we I check if a option is set twice... And get awkward checks for combinations that the assertion library doesn't provide for (null or callable or string).
Private setters are more expressive and guard values in logical way.
Bahavior; With private setters it is still immutable. 

The named constructor does seem helpful, but I think it would be more elegant to keep the setters in this case?

**Comment by bakura10** [comment]
Actually as suggested @danizord an alternative to remove teh dependency to Assertion would be to do that:

```php
private function construct(
   bool $param1,
   int $param2
) {
   $this->param1 = $param1;
   $this->param2 = $param2;
}

static public function fromArray(array $options)
{
   return new self(
     $options['param1'] ?? true,
     $options['param2'] ?? 400
   );
}
```

Therefore you keep the built-in typehinting and provide true immutability. What do you think?

**Comment by basz** [comment]
I think that's better - (until (many) more options are added)

still don't see the 'true immutability' argument. private setters provide true mutability...

I will refactor...

**Comment by danizord** [comment]
@basz the point of using constructor instead of private setters is to ensure that an object is always in consistent state after construction.

**Comment by basz** [comment]
@danizord i get that, just saying that a mega constructor is not always nice.

i tend to create ValueObjects with the following method

static withSomething : returns mutated copy)
private setSomething : from (named) constructors, may do guarding
public getSomething : get value

**Comment by danizord** [comment]
I see, so it's a matter of preference. I think setters are more verbose and PHPStorm generates the constructor automatically anyway :P

**Pull request merged**